working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+it does _not_ use a wrapper object, it enforces the desired discipline
+statically.
+
+Like extension methods, extension types are static in nature: An extension
+type declaration may declare some type parameters (just like the current
+extension declarations). The type parameters will be bound to a type which
+is determined by the static type of the receiver. Similarly, like extension
+methods, members of an extension type are resolved statically, i.e., if
+`tiny.leaves` is an invocation of an extension type getter `leaves`, then
+the declaration named `leaves` whose body is executed is determined at
+compile-time. There is no support for late binding of an extension method,
+and hence there is no notion of overriding. In return for this lack of
+expressive power, we get improved performance.
+
+Here is another example. It illustrates the fact that an extension type `E`
+with on-type `T` introduces a type `E` which is a supertype of `T`. This
+makes it possible to assign an expression of type `T` to a variable of type
+`E`. This corresponds to "entering" the extension type (accepting the
+specific discipline associated with `E`). Conversely, a cast from `E` to
+`T` is a downcast, and hence it must be written explicitly. This cast
+corresponds to "exiting" the extension type (allowing for violations of the
+discipline associated with `E`), and the fact that the cast must be written
+explicitly helps developers maintaining the discipline as intended, rather
+than dropping out of it by accident, silently.
+
+```dart
+extension type ListSize<X> on List<X> {
+  int get size => length;
+  X front() => this[0];
+}
+
+void main() {
+  ListSize<String> xs = <String>['Hello']; // OK, upcast.
+  print(xs); // OK, `toString()` available on Object?.
+  print("Size: ${xs.size}. Front: ${xs.front()}"); // OK.
+  xs[0]; // Error, `operator []` not a member of `ListSize`.
+
+  List<ListSize<String>> ys = [xs]; // OK.
+  List<List<String>> ys2 = ys; // Error, downcast.
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <identifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)* 
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::= 
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*Note that `<type>` derives `<typeIdentifier>`, which makes `<identifier>`
```

Comment:
Point to the derivation you're discussing here (extensionShowHideElement).  I was very confused by this comment for a while, thinking it referred to the first production in the list above.